File: C_Sharp/SuffixArray/InvertedSuffixArray/BinarySearch/InvertedSuffixArrayInsertSearchStrategy.cs
using SuffixArray.BinarySearch;

namespace SuffixArray.InvertedSuffixArray.BinarySearch
{
    public class InvertedSuffixArrayInsertSearchStrategy : IBinarySearchStrategy
    {
        private readonly AbstractInvertedSuffixArray array;
        private readonly int posLongestSuffix;
        private readonly char ch;

        public InvertedSuffixArrayInsertSearchStrategy(AbstractInvertedSuffixArray array, int posLongestSuffix, char ch)
        {
            this.array = array;
            this.posLongestSuffix = posLongestSuffix;
            this.ch = ch;
        }

        public bool Condition(int pos)
        {
            int idx = array.Array[pos];

            char fstCh = array.String[array.StringLength - 1 - idx];
            if (fstCh == ch)
            {
                if (idx == 0)
                {
                    return true;
                }

                int inverse = array.Array.GetInverse(idx - 1);

                return posLongestSuffix > inverse;
            }

            return fstCh < ch;
        }
    }
}

File: cpp/sa/InvertedSuffixArrayInsertSearchStrategy.cpp
#include "AbstractBinarySearchStrategy.h"
#include "InvertedSuffixArray.h"
#include "InvertedSuffixArrayInsertSearchStrategy.h"

InvertedSuffixArrayInsertSearchStrategy::InvertedSuffixArrayInsertSearchStrategy(InvertedSuffixArray* array, int posLongestSuffix, char ch)
{
  this->array = array;
  this->posLongestSuffix = posLongestSuffix;
  this->ch = ch;
}

bool InvertedSuffixArrayInsertSearchStrategy::Condition(int pos)
{
  int idx = array->array->get_elem(pos);
  
  char fstCh = array->str->getElem(array->str->Length() - 1 - idx);
  if (fstCh == ch)
    {
      if (idx == 0)
	{
	  return 1;
	}
      
      int inverse = array->array->GetInverse(idx - 1);
      
      return posLongestSuffix > inverse;
    }
  
  return fstCh < ch;
}
